### PR TITLE
rabbit-examples

### PR DESCRIPTION
**Pull request by joshuacox**
from the readme:

I needed to test a rabbitMQ with some basics so I forked the official tutorials and added a Dockerfile
[here](https://github.com/joshuacox/rabbitmq-tutorials/tree/jsdockerfile/javascript-nodejs)

[dockerhub here](https://hub.docker.com/r/joshuacox/rabbitmq-tutorials/)

In that fork, there is a `Makefile` that can run through the proof of
concept.  `make` will start a rabbitmq container, then start all the
listeners as deamons, and the emitters to run once.  The emitters are
ephemeral and will go away once their message is sent or they error out.

### CODE REVIEW

File: templates/rabbitmq-examples/0/docker-compose.yml.tpl [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+services:
+{{- if eq .Values.rabbitmq_link ""}}
+  rabbit:
+    hostname: rabbit
+    image: rabbitmq:3
+  send:
+    image: joshuacox/rabbitmq-tutorials
+    labels:
+      io.rancher.container.hostname_override: container_name
+      io.rancher.scheduler.affinity:host_label: ${host_label}
```
Comment (rawmind0) [inline templates/rabbitmq-examples/0/docker-compose.yml.tpl]: Not ${host_label} variable defined....
```
-      io.rancher.scheduler.affinity:host_label: ${host_label}
```

File: templates/rabbitmq-examples/0/docker-compose.yml.tpl [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+  receive:
+    command: receive
+{{- if ne .Values.rabbitmq_link ""}}
+    external_links:
+      - ${rabbitmq_link}:rabbit
+{{- end}}
+    image: joshuacox/rabbitmq-tutorials
+    labels:
+      io.rancher.container.hostname_override: container_name
+      io.rancher.scheduler.affinity:host_label: ${host_label}
```
Comment (rawmind0) [inline templates/rabbitmq-examples/0/docker-compose.yml.tpl]: Not ${host_label} variable defined....
```
-      io.rancher.scheduler.affinity:host_label: ${host_label}
```

File: templates/rabbitmq-examples/0/docker-compose.yml.tpl [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+    environment:
+      - AMQ_HOST=rabbit
+  send:
+    image: joshuacox/rabbitmq-tutorials
+    labels:
+      io.rancher.container.hostname_override: container_name
+      io.rancher.scheduler.affinity:host_label: ${host_label}
```
Comment (rawmind0) [inline templates/rabbitmq-examples/0/docker-compose.yml.tpl]: Not ${host_label} variable defined....
```
-      io.rancher.scheduler.affinity:host_label: ${host_label}
```

File: templates/rabbitmq-examples/0/docker-compose.yml.tpl [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+{{- if ne .Values.rabbitmq_link ""}}
+    external_links:
+      - ${rabbitmq_link}:rabbit
+{{- end}}
+
+  worker:
+    image: joshuacox/rabbitmq-tutorials
+    labels:
+      io.rancher.container.hostname_override: container_name
+      io.rancher.scheduler.affinity:host_label: ${host_label}
```
Comment (rawmind0) [inline templates/rabbitmq-examples/0/docker-compose.yml.tpl]: Not ${host_label} variable defined....
```
-      io.rancher.scheduler.affinity:host_label: ${host_label}
```

File: templates/rabbitmq-examples/0/docker-compose.yml.tpl [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+    command: worker
+{{- if ne .Values.rabbitmq_link ""}}
+    external_links:
+      - ${rabbitmq_link}:rabbit
+{{- end}}
+  newtask:
+    image: joshuacox/rabbitmq-tutorials
+    labels:
+      io.rancher.container.hostname_override: container_name
+      io.rancher.scheduler.affinity:host_label: ${host_label}
```
Comment (rawmind0) [inline templates/rabbitmq-examples/0/docker-compose.yml.tpl]: Not ${host_label} variable defined....
```
-      io.rancher.scheduler.affinity:host_label: ${host_label}
```

File: templates/rabbitmq-examples/0/docker-compose.yml.tpl [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+{{- if ne .Values.rabbitmq_link ""}}
+    external_links:
+      - ${rabbitmq_link}:rabbit
+{{- end}}
+
+  rpcserver:
+    image: joshuacox/rabbitmq-tutorials
+    labels:
+      io.rancher.container.hostname_override: container_name
+      io.rancher.scheduler.affinity:host_label: ${host_label}
```
Comment (rawmind0) [inline templates/rabbitmq-examples/0/docker-compose.yml.tpl]: Not ${host_label} variable defined....
```
-      io.rancher.scheduler.affinity:host_label: ${host_label}
```

File: templates/rabbitmq-examples/0/docker-compose.yml.tpl [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+    command: rpc_server
+{{- if ne .Values.rabbitmq_link ""}}
+    external_links:
+      - ${rabbitmq_link}:rabbit
+{{- end}}
+  rpcclient:
+    image: joshuacox/rabbitmq-tutorials
+    labels:
+      io.rancher.container.hostname_override: container_name
+      io.rancher.scheduler.affinity:host_label: ${host_label}
```
Comment (rawmind0) [inline templates/rabbitmq-examples/0/docker-compose.yml.tpl]: Not ${host_label} variable defined....
```
-      io.rancher.scheduler.affinity:host_label: ${host_label}
```

File: templates/rabbitmq-examples/0/docker-compose.yml.tpl [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+{{- if ne .Values.rabbitmq_link ""}}
+    external_links:
+      - ${rabbitmq_link}:rabbit
+{{- end}}
+
+  receivelogs:
+    image: joshuacox/rabbitmq-tutorials
+    labels:
+      io.rancher.container.hostname_override: container_name
+      io.rancher.scheduler.affinity:host_label: ${host_label}
```
Comment (rawmind0) [inline templates/rabbitmq-examples/0/docker-compose.yml.tpl]: Not ${host_label} variable defined....
```
-      io.rancher.scheduler.affinity:host_label: ${host_label}
```

File: templates/rabbitmq-examples/0/docker-compose.yml.tpl [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+    command: receive_logs
+{{- if ne .Values.rabbitmq_link ""}}
+    external_links:
+      - ${rabbitmq_link}:rabbit
+{{- end}}
+  emitlog:
+    image: joshuacox/rabbitmq-tutorials
+    labels:
+      io.rancher.container.hostname_override: container_name
+      io.rancher.scheduler.affinity:host_label: ${host_label}
```
Comment (rawmind0) [inline templates/rabbitmq-examples/0/docker-compose.yml.tpl]: Not ${host_label} variable defined....
```
-      io.rancher.scheduler.affinity:host_label: ${host_label}
```

File: templates/rabbitmq-examples/0/docker-compose.yml.tpl [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+{{- if ne .Values.rabbitmq_link ""}}
+    external_links:
+      - ${rabbitmq_link}:rabbit
+{{- end}}
+
+  receivelogsdirect:
+    image: joshuacox/rabbitmq-tutorials
+    labels:
+      io.rancher.container.hostname_override: container_name
+      io.rancher.scheduler.affinity:host_label: ${host_label}
```
Comment (rawmind0) [inline templates/rabbitmq-examples/0/docker-compose.yml.tpl]: Not ${host_label} variable defined....
```
-      io.rancher.scheduler.affinity:host_label: ${host_label}
```

File: templates/rabbitmq-examples/0/docker-compose.yml.tpl [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+    command: receive_logs_direct info
+{{- if ne .Values.rabbitmq_link ""}}
+    external_links:
+      - ${rabbitmq_link}:rabbit
+{{- end}}
+  emitlogdirect:
+    image: joshuacox/rabbitmq-tutorials
+    labels:
+      io.rancher.container.hostname_override: container_name
+      io.rancher.scheduler.affinity:host_label: ${host_label}
```
Comment (rawmind0) [inline templates/rabbitmq-examples/0/docker-compose.yml.tpl]: Not ${host_label} variable defined....
```
-      io.rancher.scheduler.affinity:host_label: ${host_label}
```

File: templates/rabbitmq-examples/0/docker-compose.yml.tpl [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+    command: receive_logs_topic anonymous.info
+{{- if ne .Values.rabbitmq_link ""}}
+    external_links:
+      - ${rabbitmq_link}:rabbit
+{{- end}}
+  emitlogtopic:
+    image: joshuacox/rabbitmq-tutorials
+    labels:
+      io.rancher.container.hostname_override: container_name
+      io.rancher.scheduler.affinity:host_label: ${host_label}
```
Comment (rawmind0) [inline templates/rabbitmq-examples/0/docker-compose.yml.tpl]: Not ${host_label} variable defined....
```
-      io.rancher.scheduler.affinity:host_label: ${host_label}
```

File: templates/rabbitmq-examples/0/docker-compose.yml.tpl [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+{{- if ne .Values.rabbitmq_link ""}}
+    external_links:
+      - ${rabbitmq_link}:rabbit
+{{- end}}
+
+  receivelogstopic:
+    image: joshuacox/rabbitmq-tutorials
+    labels:
+      io.rancher.container.hostname_override: container_name
+      io.rancher.scheduler.affinity:host_label: ${host_label}
```
Comment (rawmind0) [inline templates/rabbitmq-examples/0/docker-compose.yml.tpl]: Not ${host_label} variable defined....
```
-      io.rancher.scheduler.affinity:host_label: ${host_label}
```

File: templates/rabbitmq-examples/0/docker-compose.yml.tpl [hunk 14]
```diff
@@ -0,0 +1,171 @@
+version: '2'
+services:
+{{- if eq .Values.rabbitmq_link ""}}
+  rabbit:
+    hostname: rabbit
+    image: rabbitmq:3
+  send:
```
Comment (rawmind0) [inline templates/rabbitmq-examples/0/docker-compose.yml.tpl]: ```send``` definition duplicate??

File: templates/rabbitmq-examples/0/rancher-compose.yml [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+  version: 3.6-rancher1
+  description: |
+    tutorial storage examples
+  questions:
+    - variable: rabbitmq_link
+      description: "rabbitmq service link"
+      label: "rabbitmq stack/service"
+      default: ""
+      required: false
+      type: "service"
```
Comment (rawmind0) [inline templates/rabbitmq-examples/0/rancher-compose.yml]: Need a host_label question...
```
+
+    - variable: host_label
+      label: "Host with Label to put RabbitMQ_Examples on"
+      description: |
+        Host label to use as RabbitMQ_Examples 'value' tag.
+        Example: 'rabbitmq=true'
+      required: false
+      default: ""
+      type: "string"
```

**Comment by rawmind0** [comment]
Thanks @joshuacox. ..

You missed to define host_label question at rancher-compose.  ;)

**Comment by joshuacox** [comment]
I'm not even sure how that slipped through the cracks somehow I didn't commit that.  I think it's all good now.

File: templates/rabbitmq-examples/0/docker-compose.yml.tpl [hunk 16]
```diff
@@ -0,0 +1,186 @@
...
+{{- if eq .Values.rabbitmq_link ""}}
+  rabbit:
+    hostname: rabbit
+    image: rabbitmq:3
+{{- end}}
+  receive:
+    command: receive
+{{- if ne .Values.rabbitmq_link ""}}
+    external_links:
+      - ${rabbitmq_link}:rabbit
```
Comment (rawmind0) [inline templates/rabbitmq-examples/0/docker-compose.yml.tpl]: Redefine environment depending if you set or not $rabbitmq_link 
```
+    environment:
+      - AMQ_HOST=rabbit.rancher.internal
+{{- else}}
+    links:
+      - rabbit
+    environment:
+      - AMQ_HOST=rabbit
```

File: templates/rabbitmq-examples/0/docker-compose.yml.tpl [hunk 17]
```diff
@@ -0,0 +1,186 @@
...
+{{- end}}
+    image: joshuacox/rabbitmq-tutorials
+    labels:
+      io.rancher.container.hostname_override: container_name
+{{- if ne .Values.host_label ""}}
+      io.rancher.scheduler.affinity:host_label: ${host_label}
+{{- end}}
+    environment:
+      - AMQ_HOST=rabbit
```
Comment (rawmind0) [inline templates/rabbitmq-examples/0/docker-compose.yml.tpl]: Remove environment definition...
```
-    environment:
-      - AMQ_HOST=rabbit
```

**Comment by rawmind0** [comment]
Hey @joshuacox ...

The package is not working well. If you don't set any rabbitmq link, a rabbit is deployed and it's working, but if you set one, it's not working. 
I guess your images are based in alpine. Alpine has a dns resolv when you have more that one search domain, just try to resolv with the first one. External link to rabbit is created, but is not resolved inside your dockers. A redefinition of env var ```AMQ_HOST=rabbit```  is needed. Also, add a link to rabbit when no link is more than recommended just for manage deployment order. Take a look to the change request. The same has to be done for every service. ;)

Otherwise, i saw that the package is a kind of units or blocks that by one hand write topics to rabbitmq (run-once:true) and by other hand read that data. What do you think that write packages would be deployed as a sidekick of read packages?? 
For example, ```send``` would be a sidekick of ```receive``` 
I try to deploy like that and imho seems more useful to have the writer as a sidekick of the reader. Take a look ```rabbitmq-examples``` package at.... https://github.com/rawmind0/service-catalog
:)

**Comment by joshuacox** [comment]
I think that fixed the external links completely, thanks for the feedback!

**Comment by joshuacox** [comment]
And ya, having it done with sidekicks makes more sense

**Comment by rawmind0** [comment]
Cool, thanks... 

LGTM..

**Pull request merged**